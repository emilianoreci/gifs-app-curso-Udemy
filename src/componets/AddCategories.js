//  @Package
import React, { useState } from "react";
import PropTypes from "prop-types";

export const AddCategories = ({ setCategories }) => {
  const [inputValue, setInputValue] = useState("");

  const handleInputChange = (e) => {
    setInputValue(e.target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    //console.log("handleSubmit--llamado, ", inputValue); para el test

    //En vez de recibir las categories por props
    //setCategories([...categories, inputValue]);
    if (inputValue.trim().length > 2) {
      setCategories((cate) => [inputValue, ...cate]); //es un set de useState()
      setInputValue("");
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <h2>AddCategories</h2>
      <p>{inputValue}</p>
      <input type="text" value={inputValue} onChange={handleInputChange} />
    </form>
  );
};

AddCategories.propTypes = {
  setCategories: PropTypes.func.isRequired,
};
