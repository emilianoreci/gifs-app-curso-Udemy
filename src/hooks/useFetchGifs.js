//  @Package
import { useState, useEffect } from "react";

//  @Own
import { getGifs } from "../helpers/getGifs";

export const useFetchGifs = (category) => {
  const [state, setState] = useState({
    loading: true,
    data: [],
  });

  useEffect(() => {
    getGifs(category).then((gifs) =>
      setState({
        loading: false,
        data: gifs,
      })
    );
  }, [category]);
  return state;
};
