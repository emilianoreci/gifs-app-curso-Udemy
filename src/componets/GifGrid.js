//  @Package
import React from "react";
import { PropTypes } from "prop-types";

//  @Own
import { GifGridItem } from "./GifGridItem";
import { useFetchGifs } from "../hooks/useFetchGifs";

export const GifGrid = ({ category }) => {
  //hago destructuring y le cambio el nombre de data a images para usarlo.
  const { data: images, loading } = useFetchGifs(category);
  return (
    <>
      <h2 className="animate__animated animate__fadeIn">{category}</h2>
      <div className="card-grid">
        {loading && <p>cargando...</p>}
        {images.map((eachElement) => (
          <GifGridItem key={eachElement.id} {...eachElement}></GifGridItem>
          /* <GifGridItem key={eachElement.id} img={eachElement}></GifGridItem> */
        ))}
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
};
