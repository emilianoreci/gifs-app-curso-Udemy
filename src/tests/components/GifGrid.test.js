//  @Package
import React from "react";
import "@testing-library/jest-dom";
import { shallow } from "enzyme";

//  @Own
import { GifGrid } from "../../componets/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";

//falsea el uso de la funcion useFetchGifs
jest.mock("../../hooks/useFetchGifs");

describe("Prueba al archivo GifGrid", () => {
  const category = "batman";

  test("Se debe mostrar correctamente el componente <GifGrid", () => {
    //falsea el valor que retorna la fn, cuando se carga el componente por 1Â° vez.
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true,
    });

    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de mostrarse los items cuando se cargan las imagenes con useFetchGifs", () => {
    const gifs = [
      {
        id: "ab123",
        url: "http://localhost/algo/ray.jpg",
        title: "Algun titulo",
      },
      {
        id: "225",
        url: "http://localhost/algo/otro.jpg",
        title: "titulo 2....",
      },
    ];

    //simula que recibio un gif
    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false,
    });

    const wrapper = shallow(<GifGrid category={category} />);

    expect(wrapper).toMatchSnapshot();

    //como el loading esta en false, no deberia existir una <p/>
    expect(wrapper.find("p").exists()).toBe(false);

    //espera que haya tantos GifGridItem como la longitud que tenga el array gifs.
    expect(wrapper.find("GifGridItem").length).toBe(gifs.length);
  });
});
