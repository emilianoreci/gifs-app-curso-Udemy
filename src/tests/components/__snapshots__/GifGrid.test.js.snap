// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba al archivo GifGrid Se debe mostrar correctamente el componente <GifGrid 1`] = `
<Fragment>
  <h2
    className="animate__animated animate__fadeIn"
  >
    batman
  </h2>
  <div
    className="card-grid"
  >
    <p>
      cargando...
    </p>
  </div>
</Fragment>
`;

exports[`Prueba al archivo GifGrid debe de mostrarse los items cuando se cargan las imagenes con useFetchGifs 1`] = `
<Fragment>
  <h2
    className="animate__animated animate__fadeIn"
  >
    batman
  </h2>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ab123"
      key="ab123"
      title="Algun titulo"
      url="http://localhost/algo/ray.jpg"
    />
    <GifGridItem
      id="225"
      key="225"
      title="titulo 2...."
      url="http://localhost/algo/otro.jpg"
    />
  </div>
</Fragment>
`;
